### PR TITLE
Forwarding Server

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

To support recording sessions with OpenSSH nodes Teleport needs to create a in-memory SSH server in the proxy that forwards connections to a remote host. This PR creates this server but does not use it yet.

**Implementation**

* Consolidate handlers between the regular and forwarding server further in `lib/srv/authhandlers.go`.
* Created new forwarding server in `lib/srv/forward/sshserver.go`.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/1481
Fixes https://github.com/gravitational/teleport/issues/1486
Fixes https://github.com/gravitational/teleport/issues/1487

### CODE REVIEW

File: lib/srv/exec.go [hunk 1]
```diff
@@ -390,6 +387,85 @@ func collectLocalStatus(cmd *exec.Cmd, err error) (*ExecResult, error) {
...
+	e.command = command
+}
+
+// Start launches the given command returns (nil, nil) if successful.
+// ExecResult is only used to communicate an error while launching.
+func (r *remoteExec) Start(ch ssh.Channel) (*ExecResult, error) {
+	r.session.Stdout = ch
+	r.session.Stderr = ch.Stderr()
+
+	inputWriter, err := r.session.StdinPipe()
```
Comment (russjones) [inline lib/srv/exec.go]: Add comments.

File: lib/srv/exec.go [hunk 2]
```diff
@@ -390,6 +387,85 @@ func collectLocalStatus(cmd *exec.Cmd, err error) (*ExecResult, error) {
...
+}
+
+// Wait will block while the command executes then return the result as well
+// as emit an event to the Audit Log.
+func (r *remoteExec) Wait() (*ExecResult, error) {
+	// block until the remote command has finished execution
+	err := r.session.Wait()
+
+	// figure out if the command successfully exited or if it exited in failure
+	execResult, err := r.collectRemoteStatus(err)
```
Comment (russjones) [inline lib/srv/exec.go]: `execResult, err := r.collectRemoteStatus(r.session.Wait())`

File: lib/srv/exec.go [hunk 3]
```diff
@@ -390,6 +387,85 @@ func collectLocalStatus(cmd *exec.Cmd, err error) (*ExecResult, error) {
...
+
+	// emit the result of execution to the audit log
+	emitExecAuditEvent(r.ctx, r.command, execResult, err)
+
+	return execResult, trace.Wrap(err)
+}
+
+func (r *remoteExec) collectRemoteStatus(err error) (*ExecResult, error) {
+	if err != nil {
```
Comment (russjones) [inline lib/srv/exec.go]: Shift left.

File: lib/srv/forward/remote.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/gravitational/trace"
+)
+
+// newRemoteSession will create and return a *ssh.Client and *ssh.Session
+// with a remote host.
+func newRemoteSession(dstAddr string, systemLogin string, userAgent agent.Agent, authHandlers *srv.AuthHandlers) (*ssh.Client, *ssh.Session, error) {
+	// the proxy will use the agent that has been forwarded to it as the auth
+	// method when connecting to the remote host
+	if userAgent == nil {
+		return nil, nil, trace.AccessDenied("agent must be forwarded to proxy")
```
Comment (russjones) [inline lib/srv/forward/remote.go]: @kontsevoy What do you think, can we have a better error here?

File: lib/srv/forward/remote.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	clientConfig := &ssh.ClientConfig{
+		User: systemLogin,
+		Auth: []ssh.AuthMethod{
+			authMethod,
+		},
+		HostKeyCallback: authHandlers.HostKeyAuth,
+		Timeout:         defaults.DefaultDialTimeout,
+	}
+
+	// dial with a timeout
```
Comment (russjones) [inline lib/srv/forward/remote.go]: Update comment.

File: lib/srv/forward/remote.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return client, session, nil
+}
+
+// dialTimeout will both Dial (with a timeout) as well as place a timeout on
+// read/write on the underlying net.Conn.
+func dialTimeout(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
+	conn, err := net.DialTimeout(network, addr, config.Timeout)
```
Comment (russjones) [inline lib/srv/forward/remote.go]: Use `utils.DialWithDeadline`.

File: lib/srv/forward/sshserver.go [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+//   if err != nil {
+//   	return nil, trace.Wrap(err)
+//   }
+type Server struct {
+	log *log.Entry
+
+	clientConn net.Conn
+	serverConn net.Conn
+
+	agent        agent.Agent
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Comments.

File: lib/srv/forward/sshserver.go [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+	s.termHandlers = &srv.TermHandlers{
+		SessionRegistry: s.sessionRegistry,
+	}
+
+	return s, nil
+}
+
+// ID returns 0 for forwarding servers for now.
+func (s *Server) ID() string {
+	return "0"
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Return the ID of the proxy.

File: lib/srv/forward/sshserver.go [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+}
+
+// GetNamespace returns the namespace the forwarding server resides is.
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: resides in.

File: lib/srv/forward/sshserver.go [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+
+func (s *Server) Serve() {
+	config := &ssh.ServerConfig{
+		PublicKeyCallback: s.authHandlers.UserKeyAuth,
+	}
+	config.AddHostKey(s.hostCertificate)
+
+	sconn, chans, reqs, err := ssh.NewServerConn(s.serverConn, config)
+	if err != nil {
+		defer s.serverConn.Close()
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Don't defer.

File: lib/srv/forward/sshserver.go [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+
+		s.log.Errorf("Unable to create server connection: %v", err)
+		return
+	}
+
+	// build a remote session to the remote node
+	s.log.Debugf("Creating remote connection to %v@%v", sconn.User(), s.clientConn.RemoteAddr().String())
+	s.remoteClient, s.remoteSession, err = newRemoteSession(s.clientConn.RemoteAddr().String(), sconn.User(), s.agent, s.authHandlers)
+	if err != nil {
+		defer s.serverConn.Close()
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Don't defer.

File: lib/srv/forward/sshserver.go [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+
+		s.log.Errorf("Unable to create remote connection: %v", err)
+		return
+	}
+
+	// process global and channel requests
+	go s.handleConnection(sconn, chans, reqs)
+}
+
+func (s *Server) handleConnection(sconn *ssh.ServerConn, chans <-chan ssh.NewChannel, reqs <-chan *ssh.Request) {
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Pass a context here that looks for a heartbeat and will close here.

File: lib/srv/forward/sshserver.go [hunk 13]
```diff
@@ -0,0 +1,599 @@
...
+	ctx.AddCloser(s.serverConn)
+	ctx.AddCloser(s.clientConn)
+	ctx.AddCloser(s.remoteSession)
+	ctx.AddCloser(s.remoteClient)
+
+	defer ctx.Debugf("Closed direct-tcp context")
+	defer ctx.Close()
+
+	addr := fmt.Sprintf("%v:%d", req.Host, req.Port)
+	ctx.Infof("direct-tcpip channel: %#v to --> %v", req, addr)
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Debugf.

File: lib/srv/forward/sshserver.go [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+		if err != nil {
+			s.log.Infof("Unable to accept channel: %v", err)
+		}
+		go s.handleDirectTCPIPRequest(sconn, ch, req)
+	default:
+		nch.Reject(ssh.UnknownChannelType, fmt.Sprintf("unknown channel type: %v", channelType))
+	}
+}
+
+// handleDirectTCPIPRequest handles port forwarding requests.
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Users should be able to disable this.

File: lib/srv/forward/sshserver.go [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+	ctx.AddCloser(s.clientConn)
+	ctx.AddCloser(s.remoteSession)
+	ctx.AddCloser(s.remoteClient)
+
+	defer ctx.Debugf("Closed direct-tcp context")
+	defer ctx.Close()
+
+	addr := fmt.Sprintf("%v:%d", req.Host, req.Port)
+	ctx.Infof("direct-tcpip channel: %#v to --> %v", req, addr)
+	conn, err := net.Dial("tcp", addr)
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: This should be forwarded to the OpenSSH server.

File: lib/srv/forward/sshserver.go [hunk 16]
```diff
@@ -0,0 +1,599 @@
...
+		return s.handleEnv(ch, req, ctx)
+	case sshutils.SubsystemRequest:
+		return s.handleSubsystem(ch, req, ctx)
+	case sshutils.AgentForwardRequest:
+		// to maintain interoperability with OpenSSH, agent forwarding requests
+		// should never fail, all errors should be logged and we should continue
+		// processing requests.
+		err := s.handleAgentForward(ch, req, ctx)
+		if err != nil {
+			s.log.Info(err)
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Debugf.

File: lib/srv/forward/sshserver.go [hunk 17]
```diff
@@ -0,0 +1,599 @@
...
+		}
+		return nil
+	default:
+		return trace.BadParameter(
+			"%v doesn't support request type '%v'", s.Component(), req.Type)
+	}
+}
+
+func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.ServerContext) error {
+	// check if the users rbac role allows agent forwarding
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: check if the user's RBAC role allows agent forwarding

File: lib/srv/forward/sshserver.go [hunk 18]
```diff
@@ -0,0 +1,599 @@
...
+	err := s.authHandlers.CheckAgentForward(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// forward requests to agent passed in from the proxy and held in the context
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: Improve comments.

File: lib/srv/forward/subsystem.go [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+		log: log.WithFields(log.Fields{
+			trace.Component: teleport.ComponentRemoteSubsystem,
+			trace.ComponentFields: map[string]string{
+				"name": subsytemName,
+			},
+		}),
+		serverContext: serverContext,
+		subsytemName:  subsytemName,
+		ctx:           ctx,
+		errorCh:       make(chan error),
```
Comment (russjones) [inline lib/srv/forward/subsystem.go]: Should have size 3.

File: lib/srv/regular/sshserver.go [hunk 20]
```diff
@@ -723,45 +732,41 @@ func (s *Server) dispatch(ch ssh.Channel, req *ssh.Request, ctx *srv.ServerConte
...
 			return s.handleEnv(ch, req, ctx)
-		case sshutils.AgentReq:
-			// to maintain interoperability with OpenSSH, agent forwarding requests
-			// should never fail, so accept the request, do nothing, and return success
+		case sshutils.AgentForwardRequest:
+			// process agent forwarding, but we will only forward agent to proxy in
+			// recording proxy mode.
+			err := s.handleAgentForwardProxy(req, ctx)
+			if err != nil {
+				log.Info(err)
```
Comment (russjones) [inline lib/srv/regular/sshserver.go]: Debugf.

File: lib/srv/regular/sshserver.go [hunk 21]
```diff
@@ -810,30 +815,42 @@ func (s *Server) handleAgentForwardNode(ch ssh.Channel, req *ssh.Request, ctx *s
...
-		err = term.SetWinSize(*params)
-		if err != nil {
-			ctx.Error(err)
-		}
+	if clusterConfig.GetSessionRecording() != services.RecordAtProxy {
+		return trace.BadParameter("agent forwarding to proxy only supported in recording mode")
 	}
-	err = s.reg.NotifyWinChange(*params, ctx)
+
+	// check if the users rbac role allows agent forwarding
```
Comment (russjones) [inline lib/srv/regular/sshserver.go]: check if the user's RBAC role allows agent forwarding

File: lib/srv/term.go [hunk 22]
```diff
@@ -254,24 +269,216 @@ func (t *terminal) GetTerminalParams() rsession.TerminalParams {
...
+	t.ptyBuffer = &ptyBuffer{
+		r: stdout,
+		w: stdin,
+	}
+
+	// if a specific term type was not requested, then pick the default one and request a pty
+	if t.termType == "" {
+		t.termType = defaultTerm
+	}
+	if err := t.session.RequestPty(t.termType, t.params.W, t.params.H, ssh.TerminalModes{}); err != nil {
```
Comment (russjones) [inline lib/srv/term.go]: See if we get `ssh.TerminalModes` in the request.

File: lib/srv/term.go [hunk 23]
```diff
@@ -254,24 +269,216 @@ func (t *terminal) GetTerminalParams() rsession.TerminalParams {
...
+		Wpx uint32
+		Hpx uint32
+	}
+	req := windowChangeRequest{
+		W:   uint32(w),
+		H:   uint32(h),
+		Wpx: uint32(w * 8),
+		Hpx: uint32(h * 8),
+	}
+	_, err := t.session.SendRequest("window-change", false, ssh.Marshal(&req))
```
Comment (russjones) [inline lib/srv/term.go]: Constant.

File: lib/srv/forward/sshserver.go [hunk 24]
```diff
@@ -288,6 +307,43 @@ func (s *Server) handleConnection(sconn *ssh.ServerConn, chans <-chan ssh.NewCha
...
+		case <-keepAliveTick.C:
+			// send a keep alive with a timeout to not wait forever
+			ok := s.sendKeepAliveWithTimeout(defaults.ReadHeadersTimeout)
+			if ok {
+				missed = 0
+				continue
+			}
+
+			// if we miss 3 in a row the connections dead, call cancel and cleanup
+			missed = missed + 1
```
Comment (klizhentas) [inline lib/srv/forward/sshserver.go]: why not missed += 1

**Comment by russjones** [comment]
@r0mant Can you take a look at `lib/auth/init.go` I made some changes to how `ClusterID` is set.

File: lib/auth/init.go [hunk 25]
```diff
@@ -105,11 +105,15 @@ type InitConfig struct {
...
 	AuthPreference services.AuthPreference
 
-	// ClusterConfig holds cluster level configuration.
-	ClusterConfig services.ClusterConfig
-
 	// AuditLog is used for emitting events to audit log
 	AuditLog events.IAuditLog
+
+	// SessionRecording determines where the session is recorded: node, proxy, or off.
+	SessionRecording string
```
Comment (russjones) [inline lib/auth/init.go]: Move to `lib/config/configuration.go`

File: lib/services/clusterconfig.go [hunk 26]
```diff
@@ -110,20 +117,39 @@ const (
...
 	RecordOff RecordingType = "off"
 )
 
+// HostKeyCheckType holds how the recording proxy will check host keys.
+type HostKeyCheckType string
+
+const (
+	// HostKeyCheckOn is the default. The proxy will check the host key of the
+	// target node it connects to.
+	HostKeyCheckOn HostKeyCheckType = "on"
```
Comment (russjones) [inline lib/services/clusterconfig.go]: `yes`

File: lib/services/clusterconfig.go [hunk 27]
```diff
@@ -110,20 +117,39 @@ const (
...
+	// RecordAtProxy enables the recording proxy which intercepts and records
+	// all sessions.
 	RecordAtProxy RecordingType = "proxy"
 
 	// RecordOff is used to disable session recording completely.
 	RecordOff RecordingType = "off"
 )
 
+// HostKeyCheckType holds how the recording proxy will check host keys.
+type HostKeyCheckType string
```
Comment (russjones) [inline lib/services/clusterconfig.go]: Check.

File: lib/services/clusterconfig.go [hunk 28]
```diff
@@ -218,6 +264,9 @@ const ClusterConfigSpecSchemaTemplate = `{
     "session_recording": {
       "type": "string"
     },
+    "proxy_checks_host_keys": {
+      "type": "string"
+	},
```
Comment (russjones) [inline lib/services/clusterconfig.go]: indent

File: lib/srv/authhandlers.go [hunk 29]
```diff
@@ -79,6 +79,32 @@ func (h *AuthHandlers) CheckAgentForward(ctx *ServerContext) error {
 	return nil
 }
 
+// CheckPortForward checks if the certificate was signed by a valid certificate
+// authority, fetches the appropriate roles, and checks if port forwarding is
+// allowed.
+func (h *AuthHandlers) CheckPortForward(ctx *ServerContext) error {
```
Comment (russjones) [inline lib/srv/authhandlers.go]: Get CA set and RoleSet and pass identity earlier.

File: lib/srv/authhandlers.go [hunk 30]
```diff
@@ -197,8 +241,15 @@ func (h *AuthHandlers) HostKeyAuth(hostport string, remote net.Addr, key ssh.Pub
 		return nil
 	}
 
-	// take any valid public key and if we have gotten to this point we have
-	// a valid public key
+	// if we are strictly checking host keys then reject this request right away
+	if clusterConfig.GetProxyChecksHostKeys() == services.HostKeyCheckOn {
+		return trace.AccessDenied("certificate not presented")
```
Comment (russjones) [inline lib/srv/authhandlers.go]: `proxy expects certificate signed by certificate authority, got public key`

File: lib/srv/authhandlers.go [hunk 31]
```diff
@@ -197,8 +241,15 @@ func (h *AuthHandlers) HostKeyAuth(hostport string, remote net.Addr, key ssh.Pub
...
+	}
+
+	// if we are not stricting rejecting host keys, we need to log that we
+	// trusted a insecure key and then return nil
+	log.Infof("Strict host key checking disabled, allowing login without checking host key.")
```
Comment (russjones) [inline lib/srv/authhandlers.go]: `Warn`

Comment (russjones) [reply lib/srv/authhandlers.go]: Insecure warning prefix.

File: lib/srv/forward/sshserver.go [hunk 32]
```diff
@@ -423,6 +423,14 @@ func (s *Server) handleDirectTCPIPRequest(sconn *ssh.ServerConn, ch ssh.Channel,
 	defer ctx.Debugf("Closed direct-tcp context")
 	defer ctx.Close()
 
+	// check if the role allows port forwarding for this user
+	err := s.authHandlers.CheckPortForward(ctx)
+	if err != nil {
+		s.log.Infof(err.Error())
```
Comment (russjones) [inline lib/srv/forward/sshserver.go]: `Warn`

Comment (russjones) [reply lib/srv/forward/sshserver.go]: Emit Audit Log.

File: lib/srv/regular/sshserver.go [hunk 33]
```diff
@@ -621,6 +621,14 @@ func (s *Server) handleDirectTCPIPRequest(sconn *ssh.ServerConn, ch ssh.Channel,
 	defer ctx.Debugf("direct-tcp closed")
 	defer ctx.Close()
 
+	// check if the role allows port forwarding for this user
+	err := s.authHandlers.CheckPortForward(ctx)
+	if err != nil {
+		log.Info(err.Error())
```
Comment (russjones) [inline lib/srv/regular/sshserver.go]: `Warn`

File: lib/srv/authhandlers.go [hunk 34]
```diff
@@ -53,53 +53,67 @@ type AuthHandlers struct {
...
+		// emit port forward failure event
+		h.AuditLog.EmitAuditEvent(events.PortForwardEvent, events.EventFields{
+			events.PortForwardAddr:    addr,
+			events.PortForwardSuccess: false,
+			events.PortForwardErr:     systemErrorMessage,
+			events.EventLogin:         ctx.Identity.Login,
+			events.LocalAddr:          ctx.Conn.LocalAddr().String(),
+			events.RemoteAddr:         ctx.Conn.RemoteAddr().String(),
+		})
+		h.Warnf("Port forwarding request denied: %v", systemErrorMessage)
```
Comment (russjones) [inline lib/srv/authhandlers.go]: `.`

**Review by klizhentas — APPROVED** [review]

**Pull request merged**